lib/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1212 @@
+		return "failure";
+	case HAWKBIT_STATUS_FINISHED_NONE:
+		return "none";
+	default:
+		LOG_ERR("%d is invalid", (int)f);
+		return NULL;
+	}
+}
+
+static const char *hawkbit_status_execution(enum hawkbit_status_exec e)
+{
+	switch (e) {
+	case HAWKBIT_STATUS_EXEC_CLOSED:
+		return "closed";
+	case HAWKBIT_STATUS_EXEC_PROCEEDING:
+		return "proceeding";
+	case HAWKBIT_STATUS_EXEC_CANCELED:
+		return "canceled";
+	case HAWKBIT_STATUS_EXEC_SCHEDULED:
+		return "scheduled";
+	case HAWKBIT_STATUS_EXEC_REJECTED:
+		return "rejected";
+	case HAWKBIT_STATUS_EXEC_RESUMED:
+		return "resumed";
+	case HAWKBIT_STATUS_EXEC_NONE:
+		return "none";
+	default:
+		LOG_ERR("%d is invalid", (int)e);
+		return NULL;
+	}
+}
+
+/**
+ * @brief Update an ACID of a given type on flash
+ *
+ * @param type ACID type to update
+ * @param acid New ACID value
+ * @return 0 on success, negative on error.
+ */
+static int hawkbit_device_acid_update(enum hawkbit_dev_acid_t type,
+					u32_t new_value)
+{
+	struct hawkbit_device_acid device_action_id;
+	int ret;
+
+	ret = nvs_read(&fs, ADDRESS_ID, &device_action_id,
+		       sizeof(device_action_id));
+	if (ret < 0) {
+		LOG_ERR("Failed to read device action id");
+		return ret;
+	}
+
+	if (type == HAWKBIT_ACTION_ID_UPDATE) {
+		device_action_id.update = new_value;
+	} else {
+		device_action_id.current = new_value;
+	}
+
+	ret = nvs_write(&fs, ADDRESS_ID, &device_action_id,
+			sizeof(device_action_id));
+	if (ret < 0) {
+		LOG_ERR("Failed to write device id");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/*
+ * Update sleep interval, based on results from hawkbit base polling
+ * resource
+ */
+static void hawkbit_update_sleep(struct hawkbit_ctl_res *hawkbit_res)
+{
+	const char *sleep = hawkbit_res->config.polling.sleep;
+	int len;
+
+	if (strlen(sleep) != HAWKBIT_SLEEP_LENGTH) {
+		LOG_ERR("Invalid poll sleep: %s", sleep);
+	} else {
+		len = hawkbit_time2sec(sleep);
+		if (len > 0 && poll_sleep != K_SECONDS(len)) {
+			LOG_INF("New poll sleep %d seconds", len);
+			poll_sleep = K_SECONDS(len);
+		}
+	}
+}
+
+/*
+ * Find URL component for the device cancel operation and action id
+ */
+static int hawkbit_find_cancelAction_base(struct hawkbit_ctl_res *res,
+					  char *cancel_base)
+{
+	const char *href;
+	char *helper;
+	size_t len;
+
+	href = res->_links.cancelAction.href;
+	if (!href) {
+		*cancel_base = '\0';
+		return 0;
+	}
+
+	helper = strstr(href, "cancelAction/");
+	if (!helper) {
+		/* A badly formatted cancel base is a server error */
+		LOG_ERR("missing cancelBase/ in href %s", href);
+		return -EINVAL;
+	}
+
+	len = strlen(helper);
+	if (len > CANCEL_BASE_SIZE - 1) {
+		/* Lack of memory is an application error */
+		LOG_ERR("cancelBase %s is too big (len %zu, max %zu)", helper,
+			len, CANCEL_BASE_SIZE - 1);
+		return -ENOMEM;
+	}
+
+	strncpy(cancel_base, helper, CANCEL_BASE_SIZE);
+
+	helper = strtok(helper, "/");
+	helper = strtok(NULL, "/");
+	hb_context.action_id = atoi(helper);
```

Comment:
Please check the nulls as strtok() can return NULL.
Also do not use `atoi()` as that does not handle invalid values. Use `strtol()` instead and check errors